`test_drivers` fails with `SystemError: not supported by libffi` on windows w/ pypy > v7.3.9

I came back to this project after a long period away to find CI mostly passing, except for two combinations out of the test matrix:

- Windows + PyPy 3.8
- Windows + PyPy 3.9

The error message was `SystemError: not supported by libffi`.

The stack traces / pytest output was like:

```
============================= test session starts =============================
platform win32 -- Python 3.8.16[pypy-7.3.11-final], pytest-6.2.5, py-1.11.0, pluggy-1.5.0
rootdir: D:\a\purepyodbc\purepyodbc
collected 225 items

tests\test_connection.py xxsxxxsx                                        [  3%]
tests\test_cursor.py ..s...s...s...s...........sssss...............sssss [ 26%]
.......s...s...s...ss..s...s.                                            [ 39%]
tests\test_db_api_2.py ......................ssss..........sss...x..x... [ 60%]
.xxx..x....xxssssssssssx..x....xx                                        [ 75%]
tests\test_exceptions.py ..s.                                            [ 77%]
tests\test_purepyodbc.py FF                                              [ 78%]
tests\test_pyodbc_compat.py ssssssssssssssssssssssssssssssssssssssssssss [ 97%]
sssss                                                                    [100%]

================================== FAILURES ===================================
_____________________________ test_drivers[False] _____________________________

include_attributes = False

    @pytest.mark.parametrize(
        "include_attributes",
        [
            False,
            True,
        ],
    )
    def test_drivers(include_attributes: bool):
>       drivers = purepyodbc.drivers(include_attributes=include_attributes)

tests\test_purepyodbc.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
purepyodbc\__init__.py:56: in drivers
    return __driver_manager.sql_drivers(
purepyodbc\_driver_manager.py:446: in sql_drivers
    return_code = func(
C:\hostedtoolcache\windows\PyPy\3.8.16\x86\Lib\_ctypes\function.py:365: in __call__
    funcptr = self._getfuncptr(argtypes, self._restype_, thisarg, variadic_args=variadic_args)
C:\hostedtoolcache\windows\PyPy\3.8.16\x86\Lib\_ctypes\function.py:435: in _getfuncptr
    ptr = self._getfuncptr_fromaddress(argtypes, restype, variadic_args=variadic_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ctypes.CDLL.__init__.<locals>._FuncPtr object at 0x0000023d32587b50>
argtypes = [<class 'ctypes.c_void_p'>, <class 'ctypes.c_long'>, <class '_ctypes.array.c_char_Array_1000'>, <class 'ctypes.c_short'>, <class '_ctypes.pointer.LP_c_short'>, <class '_ctypes.array.c_char_Array_1000'>, ...]
restype = <class 'ctypes.c_short'>, variadic_args = 8

    def _getfuncptr_fromaddress(self, argtypes, restype, variadic_args=0):
        address = self._get_address()
        ffiargs = [argtype.get_ffi_argtype() for argtype in argtypes]
        ffires = restype.get_ffi_argtype()
>       return _ffi.FuncPtr.fromaddr(address, '', ffiargs, ffires, self._flags_,
                                     variadic_args=variadic_args)
E       SystemError: not supported by libffi

C:\hostedtoolcache\windows\PyPy\3.8.16\x86\Lib\_ctypes\function.py:425: SystemError
_____________________________ test_drivers[True] ______________________________

include_attributes = True

    @pytest.mark.parametrize(
        "include_attributes",
        [
            False,
            True,
        ],
    )
    def test_drivers(include_attributes: bool):
>       drivers = purepyodbc.drivers(include_attributes=include_attributes)

tests\test_purepyodbc.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
purepyodbc\__init__.py:56: in drivers
    return __driver_manager.sql_drivers(
purepyodbc\_driver_manager.py:446: in sql_drivers
    return_code = func(
C:\hostedtoolcache\windows\PyPy\3.8.16\x86\Lib\_ctypes\function.py:365: in __call__
    funcptr = self._getfuncptr(argtypes, self._restype_, thisarg, variadic_args=variadic_args)
C:\hostedtoolcache\windows\PyPy\3.8.16\x86\Lib\_ctypes\function.py:435: in _getfuncptr
    ptr = self._getfuncptr_fromaddress(argtypes, restype, variadic_args=variadic_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ctypes.CDLL.__init__.<locals>._FuncPtr object at 0x0000023d32587b50>
argtypes = [<class 'ctypes.c_void_p'>, <class 'ctypes.c_long'>, <class '_ctypes.array.c_char_Array_1000'>, <class 'ctypes.c_short'>, <class '_ctypes.pointer.LP_c_short'>, <class '_ctypes.array.c_char_Array_1000'>, ...]
restype = <class 'ctypes.c_short'>, variadic_args = 8

    def _getfuncptr_fromaddress(self, argtypes, restype, variadic_args=0):
        address = self._get_address()
        ffiargs = [argtype.get_ffi_argtype() for argtype in argtypes]
        ffires = restype.get_ffi_argtype()
>       return _ffi.FuncPtr.fromaddr(address, '', ffiargs, ffires, self._flags_,
                                     variadic_args=variadic_args)
E       SystemError: not supported by libffi

C:\hostedtoolcache\windows\PyPy\3.8.16\x86\Lib\_ctypes\function.py:425: SystemError
=========================== short test summary info ===========================
FAILED tests/test_purepyodbc.py::test_drivers[False] - SystemError: not suppo...
FAILED tests/test_purepyodbc.py::test_drivers[True] - SystemError: not suppor...
============ 2 failed, 115 passed, 90 skipped, 18 xfailed in 5.27s ============
Error: Process completed with exit code 1.
```

All other tests were passing, including Windows + PyPy 3.7. This led me to investigate the possibility that PyPy 3.8 and 3.9 had some kind of incompatibility with ctypes/cffi/libffi or something.

In fact it was more subtle than that. The failures were not specific to the _Python_ version, but specific to the _PyPy_ release. 

The reason PyPy 3.7 was passing was that it stopped receiving updates after v7.3.9. In CI, we were running v7.3.11 for PyPy 3.8, and v7.3.16 for PyPy 3.9.

I did a whole lot of digging around in different versions of PyPy on a Windows box, and discovered that the breaking PyPy release was [v7.3.10](https://doc.pypy.org/en/latest/release-v7.3.10.html).

Looking at the function in the stack trace, I noticed [this commit](https://github.com/pypy/pypy/commit/98056735e8bacaebd631a10e5ab30ea233aabb7b#diff-ab042b3dd16bf22b7e3d8595f182ad39d3823d76b414da7debe96081a884d16bR415-R416) which introduced a `variadic_args` argument. If I removed the `variadic_args` manually from that file in my local environment, the tests passed again.

Also interesting is that v7.3.10 saw PyPy move from cffi 1.15.0 to cffi 1.15.1, although I didn't see anything relevant in the changelogs at first glance. Maybe that has something to do with the change in behaviour though...?

I spent a whole bunch of time looking into what that meant exactly, and I never got to the bottom of it. But along the way (while I was writing up a minimal reproducible example for a prospective PyPy issue!) I just chanced upon a solution with a small change to the purepyodbc code... will raise a PR shortly.

--- a/purepyodbc/_driver_manager.py
+++ b/purepyodbc/_driver_manager.py
@@ -435,7 +435,6 @@ class DriverManager:
     ) -> list[str]:
         buffer_size = 1000
         direction = SqlFetchType.SQL_FETCH_FIRST
-        buffer_length = c_short(buffer_size)
         description_length = c_short()
         attributes_length = c_short()
         drivers = []
@@ -447,10 +446,10 @@ class DriverManager:
                 environment.handle,
                 direction.value,
                 driver_description,
-                buffer_length,
+                buffer_size,
                 byref(description_length),
                 driver_attributes,
-                buffer_length,
+                buffer_size,
                 byref(attributes_length),
             )
             self.check_success(return_code, environment)

